app/scripts/metamask-controller.js

```diff
@@ -1168,6 +1185,143 @@ export default class MetamaskController extends EventEmitter {
     }
   }
 
+  // eth_decryptMessage methods
+
+  /**
+  * Called when a dapp uses the eth_decryptMessage method.
+  *
+  * @param {Object} msgParams - The params of the message to sign & return to the Dapp.
+  * @param {Object} req - (optional) the original request, containing the origin
+  * Passed back to the requesting Dapp.
+  */
+  async newRequestDecryptMessage (msgParams, req) {
+    const promise = this.decryptMessageManager.addUnapprovedMessageAsync(msgParams, req)
+    this.sendUpdate()
+    this.opts.showUnconfirmedMessage()
+    return promise
+  }
+
+  /**
+  * Only decypt message and don't touch transaction state
+  *
+  * @param {Object} msgParams - The params of the message to decrypt.
+  * @returns {Promise<Object>} - A full state update.
+  */
+  decryptMessageInline (msgParams) {
+    log.info('MetaMaskController - decryptMessageInline')
+    const msgId = msgParams.metamaskId
+    const stripped = ethUtil.stripHexPrefix(msgParams.data)
+    const buff = Buffer.from(stripped, 'hex')
+    msgParams.data = JSON.parse(buff.toString('utf8'))
```

Comment:
Also, an error thrown here beaks our error handling - nothing gets returned to the caller, so the error is lost.

Switching this to use `async/await` would be the best way of ensuring all errors are propagated correctly.

Reply:
fixed